src/relax/op/op_common.h

```diff
@@ -33,8 +33,172 @@
+  if (input_sinfo == nullptr) {
+    ctx->ReportFatal(Diagnostic::Error(call)
+                     << "The input of unary operator should be Tensor. However, the given input is "
+                     << call->args[0]->struct_info_->GetTypeKey());
+  }
+  return GetRef<TensorStructInfo>(input_sinfo);
+}
+
+/*!
+ * \brief Get the tensor struct info of the binary operator inputs.
+ * \param call The Call of the binary tensor operators.
+ * \param ctx The error reporting context.
+ * \param op_name The name of the operator that invokes this function.
+ * \return A pair containing the input tensor struct info.
+ * \throw Throw exception if the number of inputs is not two, or any of the input struct info is not
+ * tensor struct info.
+ */
+inline std::pair<TensorStructInfo, TensorStructInfo> GetBinaryInputTensorStructInfo(
+    const Call& call, const BlockBuilder& ctx, const String& op_name) {
+  if (call->args.size() != 2) {
+    ctx->ReportFatal(Diagnostic::Error(call) << op_name << " op should have 2 arguments");
+  }
+  auto* lhs_sinfo = GetStructInfoAs<TensorStructInfoNode>(call->args[0]);
+  auto* rhs_sinfo = GetStructInfoAs<TensorStructInfoNode>(call->args[1]);
+  if (!lhs_sinfo || !rhs_sinfo) {
+    ctx->ReportFatal(Diagnostic::Error(call)
+                     << "Both lhs and rhs should be Tensor for " << op_name << " operator, but got "
+                     << call->args[0]->struct_info_->GetTypeKey() << " and "
+                     << call->args[1]->struct_info_->GetTypeKey());
+  }
+  return {GetRef<TensorStructInfo>(lhs_sinfo), GetRef<TensorStructInfo>(rhs_sinfo)};
+}
+
+/************ Op registration macro ************/
+
+/*!
+ * \brief Quick helper macro
+ * - Expose a make function to construct the node.
+ * - Register op to the registry.
+ * \param OpName The name of operator to register. The name passed in will
+ *  1. be prepended with a prefix "relax.op." as the FFI key string for the make function,
+ *  2. be prepended with a prefix "relax." as the key string in the operator registry.
+ */
+#define RELAX_REGISTER_UNARY_OP(OpName)                               \
+  TVM_REGISTER_GLOBAL("relax.op." OpName).set_body_typed([](Expr e) { \
+    static const Op& op = Op::Get("relax." OpName);                   \
+    return Call(op, {e}, Attrs(), {});                                \
+  });                                                                 \
+  TVM_REGISTER_OP("relax." OpName)                                    \
+      .set_num_inputs(1)                                              \
+      .add_argument("e", "Tensor", "The input tensor.")               \
+      .set_attr<FInferStructInfo>("FInferStructInfo", InferStructInfoUnary)
+
+inline StructInfo InferStructInfoUnary(const Call& call, const BlockBuilder& ctx) {
+  return GetUnaryInputTensorStructInfo(call, ctx, /*op_name=*/"Unary");
+}
+
+/************ Utilities ************/
+
+using FComputeOutDtype = std::function<DataType(const Call&, const BlockBuilder&,
```

Comment:
this seems to be local to InferShapeBinaryBroadcast. Also it is helpful to directly pass in FComputeOutDtype as a template parameter instead of std function. The lambda can be inlined by compiler.

Reply:
```c++
template<typename FType>
StructInfo InferStructInfoBroadcast(const Call&, BlockBuilder, FType fcompute_dtype) {
}
```